accelerator/core/mam_core.c

```diff
@@ -225,91 +189,138 @@ status_t ta_mam_init(mam_api_t *const api, const iota_config_t *const iconf, try
+status_t create_channel_fetch_all_transactions(const iota_client_service_t *const service, mam_api_t *const api,
+                                               const size_t channel_depth, tryte_t *const chid,
+                                               hash81_array_p tag_array) {
   status_t ret = SC_OK;
-  if (!service || !api || !chid || !epid || channel_depth < 1 || endpoint_depth < 1) {
-    ta_log_error("%s\n", "SC_MAM_NULL");
-    return SC_MAM_NULL;
-  }
   find_transactions_req_t *txn_req = find_transactions_req_new();
   transaction_array_t *obj_res = transaction_array_new();
-  hash81_array_p tag_array = hash81_array_new();
+
   if (mam_api_channel_create(api, channel_depth, chid) != RC_OK) {
     ret = SC_MAM_FAILED_CREATE_OR_GET_ID;
     ta_log_error("%s\n", "SC_MAM_FAILED_CREATE_OR_GET_ID");
     goto done;
   }
 
-  // Step 1. find all Messages ordinal under the current channel.
-  // TODO fix NUM_FLEX_TRITS_ADDRESS is 81 not 243
-  // TODO use `ta_find_transaction_objects(service, obj_req, obj_res)` instead of the original entangled function
   flex_trit_t chid_flex_trit[NUM_TRITS_ADDRESS];
   flex_trits_from_trytes(chid_flex_trit, NUM_TRITS_ADDRESS, chid, NUM_TRYTES_ADDRESS, NUM_TRYTES_ADDRESS);
   hash243_queue_push(&txn_req->addresses, chid_flex_trit);
-  retcode_t ret_rc = RC_OK;
-  ret_rc = iota_client_find_transaction_objects(service, txn_req, obj_res);
+  // TODO use `ta_find_transaction_objects(service, obj_req, obj_res)` instead of the original entangled function
+  retcode_t ret_rc = iota_client_find_transaction_objects(service, txn_req, obj_res);
   if (ret_rc && ret_rc != RC_NULL_PARAM) {
     ret = SC_MAM_FAILED_DESTROYED;
     ta_log_error("%s\n", "SC_MAM_FAILED_DESTROYED");
     goto done;
   }
 
   iota_transaction_t *tx = NULL;
-  TX_OBJS_FOREACH(obj_res, tx) { hash_array_push(tag_array, transaction_tag(tx)); }
+  TX_OBJS_FOREACH(obj_res, tx) {
+    if (!msg_id_exist(tag_array, transaction_tag(tx))) {
+      hash_array_push(tag_array, transaction_tag(tx));
+    }
+  }
+
+done:
+  find_transactions_req_free(&txn_req);
+  transaction_array_free(obj_res);
+  return ret;
+}
 
-  // Step 2. Calculate the smallest available ep_ord and msg_ord
-  uint32_t ch_remain_sk_local = 1 << channel_depth, ep_remain_sk_local = 1 << endpoint_depth;
+status_t ta_mam_written_msg_to_bundle(const iota_client_service_t *const service, mam_api_t *const api,
+                                      const size_t channel_depth, tryte_t *const chid, mam_psk_t_set_t psks,
+                                      mam_ntru_pk_t_set_t ntru_pks, char const *const payload,
+                                      bundle_transactions_t **bundle, tryte_t *msg_id,
+                                      mam_mss_key_status_t *key_status) {
+  status_t ret = SC_OK;
+  if (!service || !api || !chid || channel_depth < 1) {
+    ta_log_error("%s\n", "SC_MAM_NULL");
+    return SC_MAM_NULL;
```

Comment:
```
ret = SC_MAM_NULL;
ta_log_error("%s\n", ta_error_to_string(ret));
return ret;
```

Reply:
I think we can return thr status code here directly, since SC_MAM_NULL is a simpler error.
With returning status code directly, we save value allocate for one time.

Reply:
I think saving the value allocation is not really necessary since the allocation does not cause severe performance issue and it **might** be eliminated by the compiler optimization.

Anyway, at least change the `ta_log_error("%s\n", "SC_MAM_NULL")` to `ta_log_error("%s\n", "ta_error_to_string(SC_MAM_NULL)")`.

Reply:
got it